README.md
# Journalish

Take back your privacy and prevent snooping (somewhat)

### Why I made it:

One of my teachers didn't respect my privacy and went through my "rude" journal.

### How does it store notes?
I was going to use [Jsoning](https://support.glitch.com/t/jsoning-a-simple-key-value-json-based-persistent-lightweight-database/24575), but I don't like parsing, storing JSON.
It currently uses [keyv](https://www.npmjs.com/package/keyv), which allows 6 diffrent databases.

### Databases
You can place the Connection URL at config.js under `dburl`
Databases Supported (directly): 
- Redis (redis://user:pass@localhost:6379)
- Mongodb (mongodb://user:pass@localhost:27017/dbname)
- Sqlite (sqlite://path/to/database.sqlite)
- Postgresql (postgresql://user:pass@localhost:5432/dbname)
- MySQL/MariaDB (mysql://user:pass@localhost:3306/dbname)
- Memory (Leave Blank) (Only use for testing, It clears everytime you restart the app)

### API/Designing Clients
The API docs can be found under API.md
As of clients, I didn't make one yet. 

### Dependencies/Packages
- [express](https://www.npmjs.com/package/express) - To host the Journal
- [keyv](https://www.npmjs.com/package/keyv) - Store Entries
- @keyv/package - Adds support for mutiple Databases (See the `databases` section)
- [helmet](https://www.npmjs.com/package/helmet) - Hides x-powered-by to slow down hacking 
- [hpp](https://www.npmjs.com/package/hpp) - protect against HTTP Parameter Pollution attacks

CHANGELOG.md
#### 0.1.1 - Feature/Spell Check (7/5/2020)

- Added Ability to delete notes
- Fixed Spelling in README.md
- Changed Formatting in Changeog
- Recoded part of the "Create" function

#### 0.1.0 - First Relase (7/5/2020)

- First Release


server.js
const express = require("express");
const app = express();
var randomstring = require("randomstring");
const config = require("./config.js");
const Keyv = require("keyv");
const title = new Keyv(config.dburl, { namespace: config.prefix + "title" });
const body = new Keyv(config.dburl, { namespace: config.prefix + "body" });
app.disable("x-powered-by");

app.use(express.static("public"));

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/views/index.html");
});

app.post("/find", (req, res) => {
  if (!req.query.title || !req.query.body) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "All FIelds are Required.", "code": "ALL_FIELDS_REQUIRED", "result": null}'
      );
  } else if (req.query.password !== config.password) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "Wrong Password.", "code": "INCORRECT_PASSWORD", "result": null}'
      );
  } else {
    async function find (id){
     var title2 = await title.get(id); 
    var body2 = await body.get(id); 
res.json('{"title": "'+ title2 +'", "body"')
      
    }
    find(req.query.id)
  }
});
app.post("/create", (req, res) => {
  if (!req.query.title || !req.query.body) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "All FIelds are Required.", "code": "ALL_FIELDS_REQUIRED", "result": null}'
      );
  } else if (req.query.password !== config.password) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "Wrong Password.", "code": "INCORRECT_PASSWORD", "result": null}'
      );
  } else {
    res.status(200);
    var id = randomstring.generate({
      length: config.idlength || 7,
      charset: config.idtype
    });
    async function create(title, text) {
      await title.set(id, title);
      await body.set(id, text);
      return id;
    }
    var idgen = create(req.query.title, req.query.body);
    res.json(
      '{"success": true, "error": null, "code": "OK", "result": "' + id + '"}'
    );
  }
});
const listener = app.listen(process.env.PORT, () => {
  console.log("MyJournal is runnning on port " + listener.address().port);
});


index.js
const express = require("express");
const app = express();
var randomstring = require("randomstring");
const config = require("./config.js");
const Keyv = require("keyv");
const title = new Keyv(config.dburl, { namespace: config.prefix + "title" });
const body = new Keyv(config.dburl, { namespace: config.prefix + "body" });
app.disable("x-powered-by");

app.use(express.static("public"));

app.get("/", (req, res) => {
  res.sendFile(__dirname + "/views/index.html");
});

app.get("/find", (req, res) => {
  if (!req.query.title || !req.query.body) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "All FIelds are Required.", "code": "ALL_FIELDS_REQUIRED", "result": null}'
      );
  } else if (req.query.password !== config.password) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "Wrong Password.", "code": "INCORRECT_PASSWORD", "result": null}'
      );
  } else {
    async function find(id) {
      var title2 = await title.get(id);
      var body2 = await body.get(id);
      res.json(
        '{"title": "' +
          title2 +
          '", "body": "' +
          body2 +
          '", "id": "' +
          id +
          '"'
      );
    }
    find(req.query.id);
  }
});
app.post("/create", (req, res) => {
  if (!req.query.title) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "At least the title is required.", "code": "TITLE_NOT_FOUND", "result": null}'
      );
  } else if (req.query.password !== config.password) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "Wrong Password.", "code": "INCORRECT_PASSWORD", "result": null}'
      );
  } else {
    res.status(200);
    var id = randomstring.generate({
      length: config.idlength || 7,
      charset: config.idtype
    });
    async function create(title, text) {
      await title.set(id, title);
      await body.set(id, text);
    }
    create(req.query.title, req.query.body);
    res.end(
      '{"success": true, "error": null, "code": "OK", "result": "' + id + '"}'
    );
  }
});
app.delete("/delete", (req, res) => {
  if (req.query.password !== config.password) {
    res
      .status(403)
      .json(
        '{"success": false, "error": "Wrong Password.", "code": "INCORRECT_PASSWORD", "result": null}'
      );
  } else {
    res.status(200);
    async function remove() {
      await title.delete(req.query.id);
      await body.delete(req.query.id);
    }
    remove();
    res.end(
      '{"success": true, "error": null, "code": "OK", "result": "Deleted."}'
    );
  }
});
const listener = app.listen(config.port, () => {
  console.log("MyJournal is runnning on port " + listener.address().port);
});
